#Перевірка статиcтичних гіпотез. Z-test.



0. Зчитайте дані з `data.csv` у змінну data, яка має тип pandas.DataFrame. Ми будемо далі працювати з цією змінною.

In [4]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats

In [6]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [7]:
data = pd.read_csv('/content/drive/MyDrive/Data/data.csv')

Запустіть код нижче. Для коректної робити всіх подальших методів дані мають бути у вигляді одновимірного `numpy.ndarray` вектора та мати розмірність (100,). Така розмірність означає, що дані - одновимірні (колонка), якщо б розмірність була (100, 1), то дані сприймаються методами як двовимірні (таблиця), хоча для нас виглядати можуть так само.

In [8]:
data = data.data.values
data.shape

(100,)

**Завдання 1**. (7 балів)
Зчитані дані - це сума покупок в доларах наших покупців на сайті протягом липня 2025 року.

До цього місяця, ми мали продажі в середньому на чек $20.

Необхідно зʼясувати, чи вийшло у нас статистично значущо **підвищити** середній чек за липень 2025?

Аби дати відповідь - ми проведемо z-test на рівні значущості $\alpha=0.05$ двома способами. В цьому завданні ми будемо виконувати обчислення "вручну" з використанням бібліотек numpy та scipy.stats подібно до прикладу в лекції. Для цього виконайте наступні 7 кроків. Правильне виконання кожного з кроків оцінюється в 1 бал.

1. Запишіть параметри задачі у змінні Python:
    - `sample_mean` - $\bar{x}$, середнє значення в вибірці
    - `population_mean` - $\mu_0$, середнє значення в популяції (тобто те, з яким ми порівнюємо середнє вибіркове значення)
    - `population_std` - $\sigma$, вибіркове стандартне відхилення, яке ми вважаємо, що є рівним ст. відх. популяції, адже маємо достатньо велику вибірку.
    - `sample_size` - $n$, розмір вибірки
    - `alpha` - $\alpha$ рівень значущості
    
    **Увага!** Для обчислення стандартного відхилення маємо скористатись функцією `np.std(your_dataframe, ddof=1)`. Чому так - розбираємо в лекції про t-test.


2. Визначте для цієї задачі:
    - якою є гіпотеза $H_0$
    - якою є альтернативна гіпотеза $H_a$
    - з яким типом тесту ми маємо справу - лівосторонній, правосторонній чи двосторонній.

3. Обчисліть стандартну помилку SE.
4. Розрахуйте z-статистику.
5. Знайдіть критичне z-значення з допомогою бібліотеки stats.
6. Обчисліть p-value з допомогою бібліотеки stats.
7. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$. Для прийняття рішення зробіть і порівняння z-статистики з критичним значенням, і проаналізуйте p-value.

In [12]:
# Параметри задачі

sample_mean = np.mean(data)
population_mean = 20
population_std = np.std(data, ddof=1)
sample_size = len(data)
alpha = 0.05

Гіпотези

H₀: μ = 20 (середній чек дорівнює $20)

H₁: μ > 20 (середній чек вище ніж $20)

Тип тесту: правосторонній (one-tailed)

In [13]:
# Обчислюємо стандартну помилку

standard_error = population_std / np.sqrt(sample_size)
print(f"SE = {standard_error:.3f}")

SE = 0.625


In [14]:
# Розраховуємо z-статистику

z_statistic = (sample_mean - population_mean) / standard_error
print(f"z = {z_statistic:.3f}")

z = -0.995


In [15]:
# Знаходимо критичне z-значення

z_critical = stats.norm.ppf(1 - alpha)
print(f"Критичне z-значення = {z_critical:.3f}")

Критичне z-значення = 1.645


In [18]:
# Обчислюємо p-value

p_value = 1 - stats.norm.cdf(z_statistic)
print(f"p-value = {p_value:.3f}")

p-value = 0.840


In [ ]:
# Приймаємо рішення (z-статистика)

z-статистика = -0.995

Критичне значення = 1.645

Порівнюємо: -0.995 < 1.645

z-статистика менше критичного значення

***Висновок: Приймаємо H0***

In [ ]:
# Приймаємо рішення (p-value)

p-value = 0.840

Рівень значущості α = 0.05

Порівнюємо: 0.840 > 0.05

p-value значно вище за рівень значущості

***Висновок: Приймаємо H0***

**Завдання 2.** (3 бали)
Виконайте обчислення z-test з використанням бібліотеки statsmodels.

Отримайте z-статистику та р-значення.

Виведіть p-значення та зробіть висновок, чи ми маємо достатньо доказів, аби стверджувати, що середній чек зріс.

Чи зійшлись значення z-статистику та р-значення в цьому завданні з попередніми обчисленнями?

In [19]:
from statsmodels.stats import weightstats as stests

In [22]:
# Виконуємо правосторонній z-тест

z_stat_sm, p_val_sm = stests.ztest(data,
                                   value=population_mean,
                                   alternative='larger')

print(f"Результати z-тесту:")
print(f"Z-статистика: {z_stat_sm:.3f}")
print(f"P-value: {p_val_sm:.3f}")

Результати z-тесту:
Z-статистика: -0.995
P-value: 0.840


Значення z-статистики та p-value зійшлись з результатами попередніх обчислень.

***Ми не маємо достатньо доказів, щоб стверджувати зростання середнього чеку, тому приймаємо нульову гіпотезу.***